### PR TITLE
Plugins installation improvements

### PR DESCRIPTION
**Pull request by SchrodingersGat**
- Replaces https://github.com/inventree/InvenTree/pull/8502
- Closes https://github.com/inventree/InvenTree/issues/7709

### Summary

This PR improves plugin loading and install to ensure that the plugins are properly synced between different running containers, e.g. when running under docker. Ensures that the required system packages are installed before collecting plugins.

### Installation Process

- Install plugin via the web interface (frontend server)
- Plugin is installed and added to `plugins.txt`
- Activate plugin (via frontend server)
- Server offloads task (to background worker) to copy static files, etc
- Background worker checks `plugins.txt` - sees it has changed - and installs the new plugin
- Background worker then runs offline processing for the newly activated plugin

### CODE REVIEW

**Comment by SchrodingersGat** [comment]
@wolflu05 I have refactored this (again) - a much simpler approach now which ensures that the plugins get installed into all running containers as appropriate

**Comment by wolflu05** [comment]
Thanks, that looks promising, I'd like to take a closer look and test it myself, but that can take me some days...

**Comment by SchrodingersGat** [comment]
Let me know when you've had a chance to test it. This is one of the final outstanding Issues for 0.17.0

**Comment by SchrodingersGat** [comment]
@matmair if you have a chance to cast your eye over this I'd appreciate it :)

File: src/backend/InvenTree/InvenTree/tests.py [hunk 1]
```diff
@@ -1184,18 +1184,8 @@ def test_initial_install(self):
...
-
-            # Set dynamic setting to True and rerun to launch install
-            InvenTreeSetting.set_setting('PLUGIN_ON_STARTUP', True, self.user)
-            registry.reload_plugins(full_reload=True)
-
-        # Check that there was another run
-        response = registry.install_plugin_file()
-        self.assertEqual(response, True)
+        registry.reload_plugins(full_reload=True, collect=True)
+        self.assertGreater(len(settings.PLUGIN_FILE_HASH), 0)
```
Comment (matmair) [inline src/backend/InvenTree/InvenTree/tests.py:1188]: Q: Does this test ensure that the installation actually runs with this setup?

File: src/backend/InvenTree/plugin/installer.py [hunk 2]
```diff
@@ -63,39 +66,54 @@ def handle_pip_error(error, path: str) -> list:
...
+    """Return the file hash for the plugins file."""
+    import hashlib
+
+    pf = settings.PLUGIN_FILE
+
+    if not pf or not pf.exists():
+        return None
+
+    with pf.open('rb') as f:
+        return hashlib.md5(f.read()).hexdigest()
```
Comment (matmair) [inline src/backend/InvenTree/plugin/installer.py:116]: Q (non-blocking): Can we use this oportunity to move to a sha hashing function? md5 always comes up in security reviews and toolings.
```suggestion
        return hashlib.sha256(f.read()).hexdigest()
```

Comment (matmair) [inline src/backend/InvenTree/plugin/installer.py:116]: C: In the future, we could use this cool new(ish) function for efficient file digesting: https://docs.python.org/3/library/hashlib.html#hashlib.file_digest

File: src/backend/InvenTree/plugin/installer.py [hunk 3]
```diff
@@ -108,8 +126,17 @@ def install_plugins_file():
         logger.warning('Plugin file %s does not exist', str(pf))
         return
 
+    cmd = [
+        'install',
+        '--disable-pip-version-check',
+        '--no-cache-dir',
```
Comment (matmair) [inline src/backend/InvenTree/plugin/installer.py:132]: Q: Are you sure we do not want to use caching?

File: src/backend/InvenTree/plugin/installer.py [hunk 4]
```diff
@@ -203,15 +238,8 @@ def install_plugin(url=None, packagename=None, user=None, version=None):
...
-    # Check if we are running in a virtual environment
-    # For now, just log a warning
-    in_venv = sys.prefix != sys.base_prefix
-
-    if not in_venv:
-        logger.warning('InvenTree is not running in a virtual environment')
-
     # build up the command
-    install_name = ['install', '-U']
+    install_name = ['install', '-U', '--no-cache-dir', '--disable-pip-version-check']
```
Comment (matmair) [inline src/backend/InvenTree/plugin/installer.py:242]: Q: Same as above regarding caching

File: src/backend/InvenTree/plugin/plugin.py [hunk 5]
```diff
@@ -239,9 +239,10 @@ def file(cls) -> Path:
         """File that contains plugin definition."""
         return Path(inspect.getfile(cls))
 
-    def path(self) -> Path:
+    @classmethod
+    def path(cls) -> Path:
         """Path to plugins base folder."""
-        return self.file().parent
+        return cls.file().parent
```
Comment (matmair) [inline src/backend/InvenTree/plugin/plugin.py:245]: Q: Is this related or just nice to have?

**Review by matmair — APPROVED** [review]
Looks good in general; It would really be good to have https://github.com/inventree/InvenTree/issues/4150 available to know how prevalent instances without workers are

**Comment by matmair** [comment]
@SchrodingersGat I might be blind but where exactly is the offloading happening? Maybe I am too tired but I seem to miss that change in the diff

File: src/backend/InvenTree/plugin/installer.py (hunk 3 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/plugin/installer.py]: No probably not - I have reverted this now

File: src/backend/InvenTree/plugin/plugin.py (hunk 5 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/plugin/plugin.py:245]: This was needed for another approach I was trying - but in the end it was a nice addition

File: src/backend/InvenTree/plugin/installer.py (hunk 2 above)
Comment (matmair) [reply src/backend/InvenTree/plugin/installer.py]: @SchrodingersGat this is only available in 3.11 - we target 3.9

**Comment by SchrodingersGat** [comment]
> @SchrodingersGat I might be blind but where exactly is the offloading happening? Maybe I am too tired but I seem to miss that change in the diff

When you activate a plugin, the collection of static files is offloaded to the background worker:

image

For this to succeed, the background worker needs to have the same set of plugins installed!

File: src/backend/InvenTree/plugin/installer.py (hunk 2 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/plugin/installer.py]: Ah, ok good point.

**Comment by matmair** [comment]
Ah now I found the link, the check migrations function is doing the installing

**Comment by wolflu05** [comment]
Sorry, I had no time yet, I was planning to test this on the weekend. But if anything pops up then, I'll just reply here.
